File: benchmarks/rust/src/bin/editing.rs
use audio_samples_benchmarks::{Args, print_csv_header, print_csv_row, run_timed};

use audio_samples::{
    AudioEditing,
    operations::types::FadeCurve,
    sample_rate, sine_wave,
};
use std::time::Duration;

fn bench_trim(dur_s: u64, iters: usize, warmup: usize) {
    let audio = sine_wave::<f32>(440.0, Duration::from_secs(dur_s), sample_rate!(44100), 0.8);
    let dur_f = dur_s as f64;
    let n = audio.samples_per_channel().get();

    // trim() borrows &self and returns a new owned AudioSamples.
    let s = run_timed(iters, warmup, || {
        let out = audio.trim(dur_f * 0.25, dur_f * 0.75)
            .expect("trim should not fail");
        std::hint::black_box(out);
    });

    let op = format!("trim_middle_{dur_s}s");
    print_csv_row(&op, "audio_samples", dur_s, n, iters, warmup, &s);
}

fn bench_pad(dur_s: u64, iters: usize, warmup: usize) {
    let audio = sine_wave::<f32>(440.0, Duration::from_secs(dur_s), sample_rate!(44100), 0.8);
    let n = audio.samples_per_channel().get();

    // pad() borrows &self and returns a new owned AudioSamples (n + sr/2 samples).
    let s = run_timed(iters, warmup, || {
        let out = audio.pad(0.0, 0.5, 0.0f32)
            .expect("pad should not fail");
        std::hint::black_box(out);
    });

    let op = format!("pad_end_{dur_s}s");
    print_csv_row(&op, "audio_samples", dur_s, n, iters, warmup, &s);
}

fn bench_fade_in(dur_s: u64, iters: usize, warmup: usize) {
    let audio = sine_wave::<f32>(440.0, Duration::from_secs(dur_s), sample_rate!(44100), 0.8);
    let n = audio.samples_per_channel().get();

    // fade_in() mutates &mut self, so clone into a fresh copy each iteration.
    let s = run_timed(iters, warmup, || {
        let mut a = audio.clone();
        a.fade_in(0.5, FadeCurve::Linear).expect("fade_in should not fail");
        std::hint::black_box(a);
    });

    let op = format!("fade_in_{dur_s}s");
    print_csv_row(&op, "audio_samples", dur_s, n, iters, warmup, &s);
}

fn main() {
    audio_samples_benchmarks::bench_init();
    let a = Args::parse_with_defaults(1000, 100);
    print_csv_header();
    bench_trim(a.duration_s, a.iterations, a.warmup);
    bench_pad(a.duration_s, a.iterations, a.warmup);
    bench_fade_in(a.duration_s, a.iterations, a.warmup);
}

File: benchmarks/c/editing.c
/*
 * editing.c — trim, pad, and fade-in benchmarks (pure C, no FFmpeg)
 *
 * FFmpeg's filter graph has equivalents (atrim, apad, afade) but they carry
 * substantial setup overhead.  Both sides use their language's native idiom:
 * plain memory operations in C, AudioEditing methods in Rust.
 *
 * Methodology:
 *   trim / pad  — malloc + copy + free each iteration to match audio_samples,
 *                 which allocates a new owned buffer on every call.
 *   fade_in     — malloc + copy + in-place ramp multiply + free each iteration
 *                 to match audio_samples' clone-then-mutate pattern.
 *
 * noinline wrappers are used so the compiler cannot eliminate malloc/memcpy
 * sequences whose result is freed without an apparent read.
 *
 * Compile:
 *   gcc -O3 -march=native -ffast-math -std=c11 editing.c -o editing -lm
 */

#include "common.h"

/* -------------------------------------------------------------------------
 * trim_middle: extract the centre 50 % of the signal
 *   start = 0.25 × n,  length = 0.5 × n
 * ---------------------------------------------------------------------- */

__attribute__((noinline))
static float *do_trim(const float *in, int start, int len) {
    float *out = malloc((size_t)len * sizeof(float));
    memcpy(out, in + start, (size_t)len * sizeof(float));
    return out;
}

static void bench_trim(int duration_s, int iterations, int warmup, int sr) {
    int n       = sr * duration_s;
    int start   = n / 4;   /* 0.25 × n */
    int out_len = n / 2;   /* 0.75×n − 0.25×n */

    float *input = malloc((size_t)n * sizeof(float));
    generate_sine(input, n, 440.0f, 0.8f, sr);

    long long *times = malloc((size_t)iterations * sizeof(long long));

    for (int i = 0; i < warmup; i++)
        free(do_trim(input, start, out_len));

    for (int i = 0; i < iterations; i++) {
        long long t0 = get_time_ns();
        float *out = do_trim(input, start, out_len);
        times[i] = get_time_ns() - t0;
        free(out);
    }

    char op[64];
    snprintf(op, sizeof(op), "trim_middle_%ds", duration_s);
    print_stats(op, "c_native", duration_s, n, iterations, warmup, times);

    free(times); free(input);
}

/* -------------------------------------------------------------------------
 * pad_end: append 0.5 s of silence after the signal
 *   output length = n + sr/2 samples
 * ---------------------------------------------------------------------- */

__attribute__((noinline))
static float *do_pad(const float *in, int n, int pad_n) {
    int out_len = n + pad_n;
    float *out  = malloc((size_t)out_len * sizeof(float));
    memcpy(out, in, (size_t)n * sizeof(float));
    memset(out + n, 0, (size_t)pad_n * sizeof(float));
    return out;
}

static void bench_pad(int duration_s, int iterations, int warmup, int sr) {
    int n     = sr * duration_s;
    int pad_n = sr / 2;   /* 0.5 s of silence */

    float *input = malloc((size_t)n * sizeof(float));
    generate_sine(input, n, 440.0f, 0.8f, sr);

    long long *times = malloc((size_t)iterations * sizeof(long long));

    for (int i = 0; i < warmup; i++)
        free(do_pad(input, n, pad_n));

    for (int i = 0; i < iterations; i++) {
        long long t0 = get_time_ns();
        float *out = do_pad(input, n, pad_n);
        times[i] = get_time_ns() - t0;
        free(out);
    }

    char op[64];
    snprintf(op, sizeof(op), "pad_end_%ds", duration_s);
    print_stats(op, "c_native", duration_s, n, iterations, warmup, times);

    free(times); free(input);
}

/* -------------------------------------------------------------------------
 * fade_in: linear ramp over the first 0.5 s
 *   gain[i] = i / fade_samples  for i in [0, fade_samples)
 * ---------------------------------------------------------------------- */

__attribute__((noinline))
static float *do_fade_in(const float *in, int n, int fade_samples) {
    float *out = malloc((size_t)n * sizeof(float));
    memcpy(out, in, (size_t)n * sizeof(float));
    for (int j = 0; j < fade_samples; j++)
        out[j] *= (float)j / (float)fade_samples;
    return out;
}

static void bench_fade_in(int duration_s, int iterations, int warmup, int sr) {
    int n            = sr * duration_s;
    int fade_samples = sr / 2;   /* 0.5 s */
    if (fade_samples > n) fade_samples = n;

    float *input = malloc((size_t)n * sizeof(float));
    generate_sine(input, n, 440.0f, 0.8f, sr);

    long long *times = malloc((size_t)iterations * sizeof(long long));

    for (int i = 0; i < warmup; i++)
        free(do_fade_in(input, n, fade_samples));

    for (int i = 0; i < iterations; i++) {
        long long t0 = get_time_ns();
        float *out = do_fade_in(input, n, fade_samples);
        times[i] = get_time_ns() - t0;
        free(out);
    }

    char op[64];
    snprintf(op, sizeof(op), "fade_in_%ds", duration_s);
    print_stats(op, "c_native", duration_s, n, iterations, warmup, times);

    free(times); free(input);
}

/* -------------------------------------------------------------------------
 * main
 * ---------------------------------------------------------------------- */

int main(int argc, char **argv) {
    int dur    = parse_int_arg(argc, argv, "--duration",   1);
    int iters  = parse_int_arg(argc, argv, "--iterations", 1000);
    int warmup = parse_int_arg(argc, argv, "--warmup",     100);

    print_csv_header();
    bench_trim(dur, iters, warmup, 44100);
    bench_pad(dur, iters, warmup, 44100);
    bench_fade_in(dur, iters, warmup, 44100);

    return 0;
}
